pandas/core/internals.py

```diff
@@ -1476,14 +1476,16 @@ def where(self, other, cond, align=True, errors='raise',
         if transpose:
             values = values.T
 
-        other = getattr(other, 'values', other)
+        other = getattr(other, '_values', getattr(other, 'values', other))
         cond = getattr(cond, 'values', cond)
 
         # If the default broadcasting would go in the wrong direction, then
         # explicitly reshape other instead
         if getattr(other, 'ndim', 0) >= 1:
             if values.ndim - 1 == other.ndim and axis == 1:
                 other = other.reshape(tuple(other.shape + (1, )))
+            elif transpose and values.ndim == self.ndim - 1:
```

Comment:
we really need this?

Reply:
Without lines 1487-1488 `cond` and `values` have opposite dimension, which is the root cause of (the DataFrame part of) #21544. 

The issue is DatetimeTZBlock._try_coerce_args returns `values` as a 2-D array. I've updated _try_coerce_args so that `other` also is a 2-D array for consistency, which fixes part of the issue.

However, we don't run _try_coerce_args until the interior of func at 1499. Thus lines 1476-1477 (`values = values.T`) is a no-op given a DatetimeTZBlock (since at that point `values` is 1-D). `other._values` is also a 1-D DatetimeIndex. So to make `cond` consistent with the `values` and `other` we get out of _try_coerce_args it needs to be transposed.

An alternative would be changing `DatetimeTZBlock._try_coerce_args to return 1-D arrays for values and values_mask (changing 2875-2877). But if I do that it breaks setitem.